File: src/lib/server/day.ts
/**
 * Days, as the schedule counts them.
 *
 * UTC, and deliberately not the player's timezone. A board is served by being the most
 * recent one dated on or before today, so it stays up until the next one is due — which
 * means the only thing the timezone decides is what hour the swap happens, not whether
 * there is a board. Picking UTC keeps that hour the same for everyone, keeps it the same
 * as the one Cloud Scheduler and Firestore count in, and leaves nothing to get wrong
 * twice a year.
 *
 * A date is a `YYYY-MM-DD` string rather than a `Date` because that is what it is: the
 * name of a day, with no instant inside it. It also sorts and compares as a string, which
 * is what lets Firestore answer `liveOn <= today` from a single-field index.
 */

/** The day now falls on. */
export function today(now: Date = new Date()): string {
	return now.toISOString().slice(0, 10);
}

/** `days` after `day` — negative to go back. */
export function shift(day: string, days: number): string {
	const at = new Date(`${day}T00:00:00Z`);
	at.setUTCDate(at.getUTCDate() + days);
	return today(at);
}

File: pipeline/connectris_pipeline/day.py
"""Days, as the schedule counts them.

The Python half of `src/lib/server/day.ts`, and it has to agree with it: the generator
writes `liveOn` and the game reads it, so a disagreement about what today is called is a
day with two boards or none.

UTC, and deliberately not anyone's local time. A board is served by being the most recent
one dated on or before today, so the timezone decides what hour the board changes, not
whether there is one. UTC keeps that hour the same for every player, the same as the one
Cloud Scheduler fires in, and free of anything that happens twice a year.

A day is a `YYYY-MM-DD` string rather than a `date` because it crosses a wire as a string,
sorts as one, and is compared as one by the Firestore index the game reads it through.
"""

from __future__ import annotations

from datetime import UTC, datetime, timedelta


def today() -> str:
    """The day now falls on."""
    return datetime.now(UTC).date().isoformat()


def shift(day: str, days: int) -> str:
    """`days` after `day` — negative to go back."""
    return (datetime.fromisoformat(day).date() + timedelta(days=days)).isoformat()
